File: problems/advent-of-code/2019/02/02/solution.js
#!/usr/bin/env node

function intComputer(data) {
    for (let ip = 0; data[ip] !== 99; ip += 4) {
        switch (data[ip]) {
            case 1:
                data[data[ip + 3]] = data[data[ip + 1]] + data[data[ip + 2]];
                break;

            case 2:
                data[data[ip + 3]] = data[data[ip + 1]] * data[data[ip + 2]];
                break;

            default:
                throw new Error('Unknown opcode ' + data[ip]);
        }
    }
    return data[0];
}

const input = process.argv[2].split(',').map(Number);

for (let noun = 0; noun < 100; ++noun) {
    for (let verb = 0; verb < 100; ++verb) {
        let data = input.slice();
        data[1] = noun;
        data[2] = verb;
        if (intComputer(data) === 19690720) {
            console.log(100 * noun + verb);
            process.exit(0);
        }
    }
}
process.exit(1);

File: problems/advent-of-code/2019/02/02/solution.rs
use std::env::args;
use std::process::exit;

fn apply<F>(data: &mut Vec<usize>, ip: usize, op: F) where F : Fn(usize, usize) -> usize {
    let x = data[ip + 3];
    data[x] = op(data[data[ip + 1]], data[data[ip + 2]]);
}

fn int_computer(mut data: Vec<usize>) -> usize {
    let mut ip: usize = 0;
    while data[ip] != 99 {
        match data[ip] {
            1 => apply(&mut data, ip, |lhs, rhs| lhs + rhs),
            2 => apply(&mut data, ip, |lhs, rhs| lhs * rhs),
            _ => panic!("Unknown opcode {}", data[ip]),
        }
        ip += 4;
    }
    data[0]
}

fn main() {
    let input: Vec<usize> = args().nth(1).unwrap().split(",").map(|s| s.parse().unwrap()).collect();
    for noun in 0..99 {
        for verb in 0..99 {
            let mut data = input.to_vec();
            data[1] = noun;
            data[2] = verb;
            if int_computer(data) == 19690720 {
                println!("{}", 100 * noun + verb);
                exit(0);
            }
        }
    }
    exit(1);
}
